monitoring/src/main/java/org/polypheny/db/monitoring/core/MonitoringQueueImpl.java

```diff
@@ -0,0 +1,206 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.polypheny.db.monitoring.core;
+
+import com.google.common.collect.Sets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Optional;
+import java.util.Queue;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import lombok.NonNull;
+import lombok.extern.slf4j.Slf4j;
+import lombok.val;
+import org.polypheny.db.config.RuntimeConfig;
+import org.polypheny.db.monitoring.events.MonitoringEvent;
+import org.polypheny.db.monitoring.persistence.MonitoringRepository;
+import org.polypheny.db.util.background.BackgroundTask;
+import org.polypheny.db.util.background.BackgroundTask.TaskSchedulingType;
+import org.polypheny.db.util.background.BackgroundTaskManager;
+
+/**
+ * MonitoringQueue implementation which stores the monitoring jobs in a
+ * concurrentQueue and will process them with a background worker task.
+ */
+@Slf4j
+public class MonitoringQueueImpl implements MonitoringQueue {
+
+    // region private fields
+
+    /**
+     * monitoring queue which will queue all the incoming jobs.
+     */
+    private final Queue<MonitoringEvent> monitoringJobQueue = new ConcurrentLinkedQueue<>();
+    private final Set<UUID> queueIds = Sets.newConcurrentHashSet();
+    private final Lock processingQueueLock = new ReentrantLock();
+    private final MonitoringRepository repository;
+
+
+    private String backgroundTaskId;
+
+    /**
+     * Processed events since restart.
+     */
+    private long processedEvents;
+
+    // endregion
+
+    // region ctors
+
+
+    /**
+     * Ctor which automatically will start the background task based on the given boolean
+     *
+     * @param startBackGroundTask Indicates whether the background task for consuming the queue will be started.
+     */
+    public MonitoringQueueImpl( boolean startBackGroundTask, @NonNull MonitoringRepository repository ) {
+        log.info( "write queue service" );
+
+        this.repository = repository;
+
+        if ( startBackGroundTask ) {
+            this.startBackgroundTask();
+        }
+    }
+
+
+    /**
+     * Ctor will automatically start the background task for consuming the queue.
+     */
+    public MonitoringQueueImpl( @NonNull MonitoringRepository repository ) {
+        this( true, repository );
+    }
+
+    // endregion
+
+    // region public methods
+
+
+    @Override
+    protected void finalize() throws Throwable {
+        super.finalize();
+        if ( backgroundTaskId != null ) {
+            BackgroundTaskManager.INSTANCE.removeBackgroundTask( backgroundTaskId );
+        }
+    }
```

Comment:
Finalizers are deprecated and should not be used.
https://wiki.sei.cmu.edu/confluence/display/java/MET12-J.+Do+not+use+finalizers